pymatgen/io/lammps/inputs.py

```diff
@@ -14,18 +17,682 @@
+                        args.append(arg)
+
+        if len(args) == 1:
+            return args[0]
+        else:
+            return args
+
+    def contains_command(self, command: str, stage_name: str | None = None) -> bool:
+        """
+        Returns whether a given command is present in the LammpsInputFile.
+        A stage name can be given; in this case the search will happen only for this stage.
+
+        Args:
+            command (str): String with the command to find in the input file (e.g., "units").
+            stage_name (str): String giving the stage name where the change should take place.
+
+        Returns:
+            True if the command is present, False is not.
+        """
+        return bool(self.get_args(command, stage_name))
+
+    def set_args(self, command: str, argument: str, stage_name: str | None = None, how: str | int | list[int] = "all"):
+        """
+        Sets the arguments for the given command to the given string.
+        If a stage name is specified, it will be replaced or set only for this stage.
+        If the command is set multiple times in the file/stage, it will be replaced based on "how":
+        either the first occurrence, all of them, or the index of the occurrence.
+
+        Args:
+            command (str): String representing the command to change, e.g., "units".
+            argument (str): String with the new value for the command, e.g., "atomic".
+            stage_name (str): String giving the stage name where the change should take place.
+            how (str or int or list): "all" for changing all occurrences of the command within the stage_name
+                or the whole input file, "first" for the first occurrence, int i for the i-th time the command
+                is present in the stage_name or the whole input file, starting at 0. Can be a list of indexes as well.
+        """
+        # Get the stages to look in
+        stages_to_look = [stage_name] if stage_name else self.stages_names
+
+        # Translates how into range of indices
+        if how == "first":
+            how = [0]
+        elif how == "all":
+            getargs = self.get_args(command, stage_name)
+            N = 1 if isinstance(getargs, str) else len(getargs)
+            how = [i for i in range(N)]
+        elif isinstance(how, int):
+            how = [how]
+
+        # Look for occurrences in the relevant stages
+        i = 0
+        for i_stage, (stage, cmd_list) in enumerate(self.list_of_commands):
+            if stage in stages_to_look:
+                for i_cmd, (cmd, _) in enumerate(cmd_list):
+                    if command == cmd:
+                        if i in how:
+                            self.list_of_commands[i_stage][1][i_cmd][1] = argument
+                        i += 1
+
+    def add_stage(
+        self,
+        command: str | list[str] | dict[str, str | float],
+        stage_name: str | None = None,
+        after_stage: str | None = None,
+    ):
```

Comment:
It was a dict in the initial implementation. We decided to get rid of dicts because we might have different stages with the same name (by mistake or not). For instance, a LammpsInputFile could consist of multiple steps including multiple thermalizations, that could all be called "Thermalization". Writing this, I realize that I should check the behavior of the methods where a stage_name can be provided by the user, I'm pretty sure I did not pay attention to that.

About your first comment, I think this may be related to the one on the example of command. I believe this method makes things easy to add a stage defined by a list of commands as a user would write them in an input file. Playing with the list_of_commands can be a bit painful because of its structure (list of lists of lists).

Reply:
> It was a dict in the initial implementation. We decided to get rid of dicts because we might have different stages with the same name (by mistake or not). For instance, a LammpsInputFile could consist of multiple steps including multiple thermalizations, that could all be called "Thermalization". 

I see. I would advocate for enforcing unique stage names (whether you stick with the list input format or not). I say this for two reasons;

1. When we do automated output parsing, it would be good to organize the results by stage name. If the names are not unique, this will be harder to do and harder to disambiguate results
2. When actually executing the calculation, I could foresee wanting to organize the output files in subdirectories named according to stage (for example, maybe we want every stage to be contained in a folder and have an output called "output.dump". This isn't possible if the stage names are not unique.


> About your first comment, I think this may be related to the one on the example of command. I believe this method makes things easy to add a stage defined by a list of commands as a user would write them in an input file. Playing with the list_of_commands can be a bit painful because of its structure (list of lists of lists).

Yes I see what you mean. You might consider using `Tuple` for each line of commands rather than `List`, since they always have exactly 2 values (e.g., command and args, each as a `str`). Then your input would look like

```
list_of_commands = [
        ["Stage 1", [(cmd1, args1), (cmd2, args2)],
        ["Stage 2", [(cmd3, args3)]
    ]
```

I personally still feel that a `dict` might make more sense for `list_of_commands`, but you are the ones who will be using this most so I'll defer to your judgment. Semantically (see also other comment), I do think this argument name is confusing, because it's really a list of _stages_ (i.e., each element in the top-level list is the stage name plus the associated commands. So I'd suggest a name like `.stages` instead.

An in-between option, which could allow you to leverage built-in python semantics a little bit more, is to structure `.stages` as a list of dict. Each dict could explicitly set the `stage_name` and the `commands`, e.g.

```
stages = [
        {"stage_name": "Stage 1", "commands": [(cmd1, args1), (cmd2, args2)]},
        {"stage_name:" "Stage 2", "commands":  [(cmd3, args3)]}
    ]
```

This `list` of `dict` structure parallels MongoDB documents and is widely utilized elsewhere in our data processing architecture. It also has an advantage of being a little more human-readable on its own. By using this format you can

1. Still access stages by positional indexing
2. Access stages by name using a list comprehension, as you already do
3. Add stages using `.stages.append({}` or at an explicit position using `.stages[1]={"stage_name":XX, "commands":[]}

You could take this a step further by making each element of `commands` a dict too, with keys `cmd` and `args` or something, but that might be overcomplicating things.

Just some food for thought!

Reply:
I really like the "in-between" solution, indeed it could make things more clear. I have to think about it a bit more, because of how LAMMPS works : for instance, you can have a given command defined multiple times with different or with the same arguments, in the same stage or not. For this reason, having a dict with commands as keys would not work (this is what was initially done as well by Manas), and why we went for lists. We have also thought about tuples, but I feel like the future way of dealing with force-field parameters might be easier with lists (maybe this is totally wrong, actually we don't know yet how we will handle that properly).

Reply:
Hi @rkingsbury, thanks again for your comments on this. I implemented this last solution in my last commit. 

About the add_stage method: one could indeed argue about its utility now. I personally like having a helper method to construct a stage using a stage name and a list of commands (multiple formats are accepted), instead of having the (slight) trouble of formatting this on my own each time I want to add a stage to an InputFile. Both are currently possible, so that offers a bit of flexibility. However, I understand the semantic issue of having an àdd_stage` that does not take a `stage` as input. Maybe I could change the method so that it accepts either a formatted stage (it would basically be the same as .append) or a stage name and list of commands: what do you think? 

Alternatively, removing the add_stage could maybe lead to errors on the user side. Let's consider a user who wants to add a command to a given stage. The user can use the add_command for this purpose. This method would have to change to allow the user to provide a stage name or not, and if the stage name that is provided is not in the list of stage names of the current InputFile, a new stage is created. Having a small typo in the provided stage name could in this case change the behavior of LAMMPS executing the input file, because a new stage would be created at the end with the provided command instead of inserting the command to the desired stage. With the add_command as is now, this error is not possible: if the stage name is not already present, an error is thrown. The stage has to be created first. 

Another possible solution would be to have add_stage only initializing the stage (empty), and add_command being able to add different formats of commands. In this case, it could be renamed initialize_stage so that the input (stage name) is not confusing. 

Let me know what you think of these possibilities.

Reply:
> About the `add_stage` method: one could indeed argue about its utility now. I personally like having a helper method to construct a stage using a `stage_name` and `commands` (multiple formats are accepted), instead of having the (slight) trouble of formatting this on my own each time I want to add a stage to an `InputFile`. Both are currently possible, so that offers a bit of flexibility (even though the append does not make sure that all stage names are unique). However, I understand the semantic issue of having an `add_stage` that does not take a `stage` as input. Maybe I could change the method so that it accepts either a formatted stage (it would basically be the same as `.append`) or a `stage_name` and `commands`: what do you think?

I see your points about the utility here, and I hadn't thought about the fact that `.append` won't enforce unique stage names. If you think this will be useful to you I think you should leave it. And with the clarifications to the attribute name (`.stages`) and more expressive syntax, I'm personally less concerned about the semantic confusion. I think it's OK as-is.


> Alternatively, removing the `add_stage` could maybe lead to errors on the user side. Let's consider a user who wants to add a command to a given stage. The user can use the `add_command` for this purpose. This method would have to be modified to allow the user to provide a `stage_name` or not, and if the `stage_name` that is provided is not in the list of stage names of the current `InputFile`, a new stage is created. Having a small typo in the provided stage name could in this case change the behavior of LAMMPS executing the input file, because a new stage would be created at the end with the provided command instead of inserting the command to the desired stage (not sure this is super clear). With the `add_command` as it is now, this error is not possible: if the `stage_name` is not already present, an error is thrown. The stage has to be created first.

Good point! And anyway as you said, having these methods does not prevent users from directly modifying the underlying `list`/`dict`, but wrapping those modifications in methods clearly does have some benefits from a validation / error checking standpoint.

> Another possible solution would be to have `add_stage` only initializing the stage (empty), and `add_command` being able to add different formats of commands. In this case, `add_stage` could be renamed `initialize_stage` so that the input (`stage_name`) is not confusing.

I'm not in favor of this; I think it will be harder to understand and more complicated than the alternative, without a clear benefit.